velox/exec/SortBuffer.cpp

```diff
@@ -0,0 +1,360 @@
+      currentMemoryUsage > spillMemoryThreshold_) {
+    const int64_t bytesToSpill =
+        currentMemoryUsage * spillConfig_->spillableReservationGrowthPct / 100;
+    auto rowsToSpill = std::max<int64_t>(
+        1, bytesToSpill / (data_->fixedRowSize() + outOfLineBytesPerRow));
+    spill(
+        std::max<int64_t>(0, currentDataRows - rowsToSpill),
+        std::max<int64_t>(
+            0, outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow)));
+    return;
+  }
+
+  // If we have enough free rows for input rows and enough variable length
+  // free space for the vector's flat size, no need for spilling.
+  if (freeRows > input->size() &&
+      (outOfLineBytes == 0 || outOfLineFreeBytes >= flatInputBytes)) {
+    return;
+  }
+
+  // For variable length data, we take the flat size of the input as the cap.
+  const int64_t estimatedIncrementalBytes =
+      data_->sizeIncrement(input->size(), outOfLineBytes ? flatInputBytes : 0);
+
+  // If the current available reservation in memory pool is 2X the
+  // estimatedIncrementalBytes, no need to spill.
+  if (pool_->availableReservation() > 2 * estimatedIncrementalBytes) {
+    return;
+  }
+
+  // Try reserving targetIncrementBytes more in memory pool, if succeed, no
+  // need to spill.
+  const auto targetIncrementBytes = std::max<int64_t>(
+      estimatedIncrementalBytes * 2,
+      currentMemoryUsage * spillConfig_->spillableReservationGrowthPct / 100);
+  if (pool_->maybeReserve(targetIncrementBytes)) {
+    return;
+  }
+
+  // NOTE: disk spilling use the system disk spilling memory pool instead of
+  // the operator memory pool.
+  const int64_t rowsToSpill = std::max<int64_t>(
+      1, targetIncrementBytes / (data_->fixedRowSize() + outOfLineBytesPerRow));
+  spill(
+      std::max<int64_t>(0, currentDataRows - rowsToSpill),
+      std::max<int64_t>(
+          0, outOfLineBytes - (rowsToSpill * outOfLineBytesPerRow)));
+}
+
+void SortBuffer::spill(int64_t targetRows, int64_t targetBytes) {
+  VELOX_CHECK_GE(targetRows, 0);
+  VELOX_CHECK_GE(targetBytes, 0);
+
+  if (spillConfig_ != nullptr && spiller_ == nullptr) {
+    spiller_ = std::make_unique<Spiller>(
+        Spiller::Type::kOrderBy,
+        data_.get(),
+        [&](folly::Range<char**> rows) { data_->eraseRows(rows); },
+        spillerStoreType_,
+        data_->keyTypes().size(),
+        sortCompareFlags_,
+        spillConfig_->filePath,
+        spillConfig_->maxFileSize,
+        spillConfig_->minSpillRunSize,
+        spillConfig_->compressionKind,
+        Spiller::pool(),
+        spillConfig_->executor);
+    VELOX_CHECK_EQ(spiller_->state().maxPartitions(), 1);
+    // Internally, It would sort and then spill.
+    spiller_->spill(targetRows, targetBytes);
+
+    // TODO: support fine-grain memory recycling after row container memory
+    // compaction support. As of now, we only support releasing all the memory
+    // when we spill everything.
+    if (targetRows == 0 || targetBytes == 0) {
```

Comment:
The condition should be
```
if ((data_->numRows() == 0) {
   data_->clear();
}
```
Let's call pool release in reclaim method in operator instead of here.